python/tests/test_topology.py

```diff
@@ -27,17 +27,76 @@
 import unittest
 import itertools
 import random
+import json
 
 import numpy as np
 import msprime
 
 import tskit
 import _tskit
+import tskit.provenance as provenance
 import tests as tests
 import tests.tsutil as tsutil
 import tests.test_wright_fisher as wf
 
 
+def slice(
+        ts, start=None, stop=None, reset_coordinates=True, simplify=True,
+        record_provenance=True):
+    """
+    A clearer but slower implementation of TreeSequence.slice() defined in trees.py
+    """
+    if start is None:
+        start = 0
+    if stop is None:
+        stop = ts.sequence_length
+
+    if start < 0 or stop <= start or stop > ts.sequence_length:
+        raise ValueError("Slice bounds must be within the existing tree sequence")
+    tables = ts.dump_tables()
+    tables.edges.clear()
+    tables.sites.clear()
+    tables.mutations.clear()
+    for edge in ts.edges():
+        if edge.right <= start or edge.left >= stop:
+            # This edge is outside the sliced area - do not include it
+            continue
+        if reset_coordinates:
+            tables.edges.add_row(
+                max(start, edge.left) - start, min(stop, edge.right) - start,
+                edge.parent, edge.child)
+        else:
+            tables.edges.add_row(
+                max(start, edge.left), min(stop, edge.right),
+                edge.parent, edge.child)
+    for site in ts.sites():
+        if start <= site.position < stop:
+            if reset_coordinates:
+                site_id = tables.sites.add_row(
+                    site.position - start, site.ancestral_state, site.metadata)
+            else:
+                site_id = tables.sites.add_row(
+                    site.position, site.ancestral_state, site.metadata)
+            for m in site.mutations:
+                tables.mutations.add_row(
+                    site_id, m.node, m.derived_state, m.parent, m.metadata)
+    if reset_coordinates:
+        tables.sequence_length = stop - start
+    if simplify:
+        tables.simplify()
+    if record_provenance:
+        # TODO add slice arguments here
+        # TODO also make sure we convert all the arguments so that they are
+        # definitely JSON encodable.
+        parameters = {
+            "command": "slice",
+            "TODO": "add slice parameters"
```

Comment:
perhaps remove these TODOs?

Reply:
I've made a start at encoding parameters at https://github.com/tskit-dev/tskit/blob/eca54e39108423038a1b8f6112dab3b20a94e280/python/tskit/trees.py#L3108 and  https://github.com/tskit-dev/tskit/blob/eca54e39108423038a1b8f6112dab3b20a94e280/python/tskit/trees.py#L3179

But it seems like this general way of extracting and encoding params to a function using the `inspect` module should be abstracted so it can be re-used.

Also, if we are to implement this for `simplify()`, where there is a similar TODO left by @jeromekelleher , we need a way to serialize np.arrays in JSON, as we can pass a numpy array as a parameter to simplify. I haven't attempted to tackle this here, as `slice` doesn't require numpy arrays.

I'm not sure if I should open this up as another PR, and if so, whether *this* PR should include it or not.

Reply:
I've worked this up into something more general in https://github.com/tskit-dev/tskit/pull/243 - although not ready for review (e.g. no tests) there should be enough there to see the general approach.

Reply:
It seems like `slice` has only a few relatively simple arguments. Why not just make a simple JSON dictionary of them for the provenance and postpone a more general solution? It is hard for me to see the right way to do it in general without working through some specific examples - as you've pointed out, there's some possibly serious issues lurking.

Reply:
Hm, I see that's more or less what you've done.

Reply:
Yes, I guess this can be marked as "resolved", on the assumption that https://github.com/tskit-dev/tskit/pull/243 (or something like it) will eventually get merged